utils/make_build.py

```diff
@@ -61,16 +62,53 @@
     'darwin': 'mac',
 }
 
-_logger = logging.getLogger (__name__)
+_logger = logging.getLogger(__name__)
+
+
+def _select_vs_generator():
+    """
+    On Windows, discover the correct generator to request from cmake.
+
+    We must first run the vswhere.exe utility to discover all of the Visual Studio installations
+    on the host. From this we then find the latest version and map this to the corresponding
+    cmake generator name.
+
+    :return: The name of a cmake generator or None if we were not able to determine the latest
+     installed version of Visual Studio.
+    """
+
+    env_name = 'ProgramFiles(x86)'
+    program_files = os.getenv(env_name)
+    if program_files is None:
+        _logger.warning('Environment variable "%s" is not defined. Cannot find Visual Studio', env_name)
+        return None
+
+    vswhere = os.path.join(program_files, 'Microsoft Visual Studio', 'Installer', 'vswhere.exe')
+    if not os.path.exists(vswhere):
+        _logger.warning('vswhere.exe is not found at "%s"', vswhere)
+        return None
+
+    _logger.info('Running vswhere at "%s"', vswhere)
+    # Ask vswhere to disclose all of the vs installations.
+    installations = json.loads(subprocess.check_output([vswhere, '-format', 'json'], stderr=subprocess.STDOUT))
+    # Reduce the installations list to just the integer major version numbers.
+    installations = (int(install['installationVersion'].split('.')[0]) for install in installations)
+    _logger.debug('Installations are: %s', ','.join((str(inst) for inst in installations)))
+    return {
+        16: 'Visual Studio 16 2019',
+        15: 'Visual Studio 15 2017'
```

Comment:
Should add **Win64**?
        16: 'Visual Studio 16 2019 **Win64**',
        15: 'Visual Studio 15 2017 **Win64**'

Reply:
What does that do?

Reply:
It's an (optional) architecture flag. You're right, this was present in the original version for VS2017. According to the cmake help, it doesn't seem to be available VS2019. It's failing on Appveyor so perhaps it's important…

Reply:
"Visual Studio 15 2017": the default target platform is Win32.
"Visual Studio 15 2017 Win64":  the target platform is x64.

The pstore bit count (`bit_count.hpp`) implementation is based on 64-bits, for example, `_BitScanForward64` and `_BitScanReverse64`. The pstore project would be built if the solution is targeted on Win32.

Reply:
> "Visual Studio 15 2017": the default target platform is Win32.
> "Visual Studio 15 2017 Win64": the target platform is x64.

Out of curiosity, where did you find this information?

Okay, so it seems that we need the “Win64” decoration on the generator name for VS2017. 

What about VS2019? According to the [VS2019 generator doc](https://cmake.org/cmake/help/latest/generator/Visual%20Studio%2016%202019.html) I simply been getting lucky because my host is 64-bit. I need to specify the `-A` switch or the build “architecture” (is there a difference between “platform” and “architecture” here?) is derived from the host OS. That'll need some further reworking of the code. 

According to the [VS2017 generator doc], I can use the same `-A` switch. 

Perhaps I should roll the `-T` wart that we’ve currently got in the code to select the 64-bit host tools into this same code.

Grumble……

Reply:
> Out of curiosity, where did you find this information?

I find it in the `Platform Selection` section (https://cmake.org/cmake/help/latest/generator/Visual%20Studio%2015%202017.html)
```

Platform Selection

The default target platform name (architecture) is Win32.

For compatibility with CMake versions prior to 3.1, one may specify a target platform name optionally at the end of the generator name. This is supported only for:

Visual Studio 15 2017 Win64
    Specify target platform x64.
```

You are right. The VS2019 generator doc didn't mention that.